nexus/types/src/deployment/blueprint_diff.rs

```diff
@@ -0,0 +1,804 @@
+                zone.zone.disposition()
+            );
+            BpSledSubtableRow::new(
+                state,
+                vec![
+                    zone.zone.kind().to_string(),
+                    zone.zone.id().to_string(),
+                    disposition,
+                    zone.zone.underlay_address().to_string(),
+                ],
+            )
+        })
+    }
+}
+
+#[derive(Debug)]
+/// Errors arising from illegally modified zone fields
+pub struct BpDiffZoneErrors {
+    pub generation_before: Generation,
+    pub generation_after: Generation,
+    pub errors: Vec<BpDiffZoneError>,
+}
+
+#[derive(Debug)]
+pub struct BpDiffZoneError {
+    pub zone_before: BlueprintOrCollectionZoneConfig,
+    pub zone_after: BlueprintOrCollectionZoneConfig,
+    pub reason: String,
+}
+
+/// All known zones across all known sleds, their various states, and errors
+#[derive(Debug, Default)]
+pub struct BpDiffZones {
+    pub added: BTreeMap<SledUuid, BpDiffZoneDetails>,
+    pub removed: BTreeMap<SledUuid, BpDiffZoneDetails>,
+    pub unchanged: BTreeMap<SledUuid, BpDiffZoneDetails>,
+    pub modified: BTreeMap<SledUuid, BpDiffZonesModified>,
+    pub errors: BTreeMap<SledUuid, BpDiffZoneErrors>,
+}
+
+impl BpDiffZones {
+    pub fn new(
+        before: BTreeMap<SledUuid, BlueprintOrCollectionZonesConfig>,
+        mut after: BTreeMap<SledUuid, BlueprintZonesConfig>,
+    ) -> Self {
+        let mut diffs = BpDiffZones::default();
+        for (sled_id, before_zones) in before {
+            let before_generation = before_zones.generation();
+            let mut removed = vec![];
+            if let Some(after_zones) = after.remove(&sled_id) {
+                let after_generation = after_zones.generation;
+                let mut unchanged = vec![];
+                let mut modified = vec![];
+                let mut errors = vec![];
+                let mut added = vec![];
+
+                // Compare `before_zones` and `after_zones` to look
+                // for additions, deletions, modifications, and errors.
+                let before_by_id: BTreeMap<_, BlueprintOrCollectionZoneConfig> =
+                    before_zones.zones().map(|z| (z.id(), z)).collect();
+                let mut after_by_id: BTreeMap<_, BlueprintZoneConfig> =
+                    after_zones.zones.into_iter().map(|z| (z.id, z)).collect();
+
+                for (zone_id, zone_before) in before_by_id {
+                    if let Some(zone_after) = after_by_id.remove(&zone_id) {
+                        // Are the zones equal?
+                        if zone_before == zone_after {
+                            unchanged.push(zone_after.into());
+                        } else {
+                            // The zones are different. They are only allowed to differ in terms
+                            // of `disposition`, otherwise we have an error.
+                            match zone_before.is_valid_modification(&zone_after)
```

Comment:
Nit - I'd expect `is_*` to return a bool. In the spirit of "parse, don't validate", could we make this `ModifiedZone`'s constructor instead? That would also centralize the decision for "what is allowed to change".

Reply:
Excellent idea. Done.